CMake 3.18 cannot configure

With CMake 3.18, libtorrent fails to configure:

```
CMake Error at cmake/Modules/GeneratePkgConfig.cmake:181 (file):
  Error evaluating generator expression:

    $<COMPILE_LANG_AND_ID:CUDA,NVIDIA>

  $<COMPILE_LANG_AND_ID:lang,id> may only be used with binary targets to
  specify include directories, compile definitions, and compile options.  It
  may not be used with the add_custom_command, add_custom_target, or
  file(GENERATE) commands.
Call Stack (most recent call first):
  CMakeLists.txt:747 (generate_and_install_pkg_config_file)
```

(on FreeBSD at least, but this looks like a generic CMake problem). There's an upstream report from me at https://gitlab.kitware.com/cmake/cmake/-/issues/21074 and a related report in a different project at https://github.com/facebook/folly/issues/1414

--- a/cmake/Modules/GeneratePkgConfig.cmake
+++ b/cmake/Modules/GeneratePkgConfig.cmake
@@ -146,10 +146,13 @@ function(generate_and_install_pkg_config_file _target _packageName)
 	set(_generate_target_dir "${CMAKE_CURRENT_BINARY_DIR}/${_target}-pkgconfig")
 	set(_pkg_config_file_template_filename "${_GeneratePkGConfigDir}/pkg-config.cmake.in")
 
-	# Avoid things in the CFLAGS -- which can come from INTERFACE_OPTIONS
-	# of dependencies -- that don't work with file(GENERATE). CMake bug 21074
-	string(REPLACE "<COMPILE_LANG_AND_ID:CUDA,NVIDIA>" "<COMPILE_LANGUAGE:CUDA>" _s "${_interface_compile_options}")
-	set(_interface_compile_options "${_s}")
+	# Since CMake 3.18 FindThreads may include a generator expression requiring a target, which gets propagated to us through INTERFACE_OPTIONS.
+	# Before CMake 3.19 there's no way to solve this in a general way, so we work around the specific case. See #4956 and CMake bug #21074.
+	if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.19)
+		set(_target_arg TARGET ${_target})
+	else()
+		string(REPLACE "<COMPILE_LANG_AND_ID:CUDA,NVIDIA>" "<COMPILE_LANGUAGE:CUDA>" _interface_compile_options "${_interface_compile_options}")
+	endif()
 
 	# put target and project properties into a file
 	configure_file("${_GeneratePkGConfigDir}/target-compile-settings.cmake.in"
@@ -159,7 +162,7 @@ function(generate_and_install_pkg_config_file _target _packageName)
 	if (NOT _isMultiConfig)
 		set(_variables_file_name "${_generate_target_dir}/compile-settings-expanded.cmake")
 
-		file(GENERATE OUTPUT "${_variables_file_name}" INPUT "${_generate_target_dir}/compile-settings.cmake")
+		file(GENERATE OUTPUT "${_variables_file_name}" INPUT "${_generate_target_dir}/compile-settings.cmake" ${_target_arg})
 
 		configure_file("${_GeneratePkGConfigDir}/generate-pkg-config.cmake.in"
 			"${_generate_target_dir}/generate-pkg-config.cmake" @ONLY)
@@ -169,7 +172,7 @@ function(generate_and_install_pkg_config_file _target _packageName)
 		foreach(cfg IN LISTS CMAKE_CONFIGURATION_TYPES)
 			set(_variables_file_name "${_generate_target_dir}/${cfg}/compile-settings-expanded.cmake")
 
-			file(GENERATE OUTPUT "${_variables_file_name}" INPUT "${_generate_target_dir}/compile-settings.cmake" CONDITION "$<CONFIG:${cfg}>")
+			file(GENERATE OUTPUT "${_variables_file_name}" INPUT "${_generate_target_dir}/compile-settings.cmake" CONDITION "$<CONFIG:${cfg}>" ${_target_arg})
 
 			configure_file("${_GeneratePkGConfigDir}/generate-pkg-config.cmake.in"
 				"${_generate_target_dir}/${cfg}/generate-pkg-config.cmake" @ONLY)

